[mobile] Allow user to configure filesystem type

Undef-a opened the pull request:
The user can now configure the target root filesystem type. Options
provided are ext4 (default), f2fs and btrfs.

OS Maintainers are required to provide their own BTRFS install script
which configures default submodules as desired.

Commit:
[mobile] Allow user to configure filesystem type

The user can now configure the target root filesystem type. Options
provided are ext4 (default), f2fs and btrfs.

OS Maintainers are required to provide their own BTRFS install script
which configures default submodules as required.

Allow disabling selection of fstype
This setting is disabled by default and defaults to ext4. Enabling it
will provide the user with a choice of ext4, f2fs and btrfs for the root
filesystem.

Add FS specific mkfs configuration options
This allows distribution maintainers to set the command/script used for
configuring the root filesystem on a per-fs level. Two examples of its
use are:
* F2FS: mkfs.f2fs uses -l for label, rather than -L like other filesystems.
* BTRFS: Distributions can provide a script which will configure submodules
         as requiref for snapshots to work correctly.

Use of QStringList rather than a Model:
We don't need the complexities of a full model here for a list of strings
and labelModel was the wrong one anyway. At least going down that path means
the back and forward buttons work.

ollieparanoid commented on modules/mobile/Config.h:
```diff
@@ -120,6 +127,9 @@ class Config : public QObject
     bool m_isFdeEnabled = false;
 
+    /* filesystem selection */
+    QString m_fstype = "ext4";
```
Let's make the default filesystem configurable (default `ext4` makes sense).

To add config options:
* add it to mobile.conf
* add it to Config.h
* add it to [Config::SetConfigurationMap](https://github.com/calamares/calamares-extensions/blob/a53f9359f42b6c301b19060f90a5d4e8bc771977/modules/mobile/Config.cpp#L18)

Example: https://github.com/calamares/calamares-extensions/commit/7f750cb1450f133c6bc505e60192804b46ff594d

---

For the fstype option in particular, it also makes sense to turn this into an enum. @adriaandegroot wrote:
> see NamedEnum.h

(I haven't used this myself yet.)

Undef-a replied:
I believe the first half of this is done. However, I'm not sure exactly how to implement NamedEnums for this. Is there an implementation somewhere you could point me to?

ollieparanoid replied:
Thanks for adding all the config options!

There is still an option missing to make the default fs type configurable.

> However, I'm not sure exactly how to implement NamedEnums for this. Is there an implementation somewhere you could point me to?

Looking at code in calamares... the welcomeq module from calmares proper has a language selection with a combo box:

* https://github.com/calamares/calamares/blob/132ff59d9c8b20147d3e16112c70998b59b05c69/src/modules/welcomeq/welcomeq.qml#L163-L171
* https://github.com/calamares/calamares/blob/132ff59d9c8b20147d3e16112c70998b59b05c69/src/modules/welcome/Config.cpp#L83-L87
* https://github.com/calamares/calamares/blob/132ff59d9c8b20147d3e16112c70998b59b05c69/src/modules/welcome/Config.h#L119

This uses a LabelModel instead of NamedEnum though, so that's probably the right one to use here.

(@adriaandegroot: or is it?)

Undef-a replied:
I've got a partial version of this, but it doesn't compile. 

```
In file included from /home/user/QubesIncoming/phone/calamares-settings-mobian/build/modules/mobile/calamares_viewmodule_mobile_autogen/mocs_compilation.cpp:2:
/home/user/QubesIncoming/phone/calamares-settings-mobian/build/modules/mobile/calamares_viewmodule_mobile_autogen/EWIEGA46WW/moc_Config.cpp: In static member function ‘static void Config::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)’:
/home/user/QubesIncoming/phone/calamares-settings-mobian/build/modules/mobile/calamares_viewmodule_mobile_autogen/EWIEGA46WW/moc_Config.cpp:275:92: error: no match for ‘operator=’ (operand types are ‘CalamaresUtils::Locale::LabelModel’ and ‘CalamaresUtils::Locale::LabelModel*’)
  275 |         case 15: *reinterpret_cast< CalamaresUtils::Locale::LabelModel*>(_v) = _t->fsmodel(); break;
      |                                                                                            ^
In file included from /home/user/QubesIncoming/phone/calamares-settings-mobian/build/modules/mobile/calamares_viewmodule_mobile_autogen/EWIEGA46WW/../../../../../modules/mobile/Config.h:6,
                 from /home/user/QubesIncoming/phone/calamares-settings-mobian/build/modules/mobile/calamares_viewmodule_mobile_autogen/EWIEGA46WW/moc_Config.cpp:10,
                 from /home/user/QubesIncoming/phone/calamares-settings-mobian/build/modules/mobile/calamares_viewmodule_mobile_autogen/mocs_compilation.cpp:2:
/usr/include/libcalamares/locale/LabelModel.h:27:17: note: candidate: ‘CalamaresUtils::Locale::LabelModel& CalamaresUtils::Locale::LabelModel::operator=(const CalamaresUtils::Locale::LabelModel&)’ (deleted)
   27 | class DLLEXPORT LabelModel : public QAbstractListModel
      |                 ^~~~~~~~~~
/usr/include/libcalamares/locale/LabelModel.h:27:17: note:   no known conversion for argument 1 from ‘CalamaresUtils::Locale::LabelModel*’ to ‘const CalamaresUtils::Locale::LabelModel&’
```

Undef-a replied:
EDIT: This feature is now working properly, both backwards and forwards. 

I have now got this working with one exception: Instead of saving the filesystem name into `config.fsType` it saves a locale string in the format `* f2fs (C)`. 

Otherwise, the default is now set on entry and (bug aside) the selected filesystem would be correctly selected if back is pressed.

ollieparanoid commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -94,7 +110,10 @@ PartitionJob::exec()
+    commands.append( {
+			{ { "mkdir", "-p", pathMount }, QString() },
+            { { "parted", devNoPart, "mkpart", "primary", "ext4", "268M", "4.9G", "-s" }, QString() },
+	} );
```
(Tabs and spaces mixed. Please make use of `.clang-format` in the git repo to automatically format your code.)

ollieparanoid commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -83,8 +85,22 @@ PartitionJob::exec()
     QString cryptDev = "/dev/mapper/" + cryptName;
     QString passwordStdin = m_password + "\n";
     QString dev = m_targetDeviceRoot;
+    QString devNoPart = dev.chopped(2);
```
We can't always chop off two characters to go from partition to the device with the partition. When I [run this in QEMU](https://wiki.postmarketos.org/wiki/On-device_installer#Running_in_QEMU), `dev` may be `/dev/vda2` for example. (opposed to `/dev/mmcblk0p2` etc, where there is always a `p` character).

But before changing this, please explain why it is needed at all. Below you are creating an additional ext4 partition. Is this an unrelated change that made it into the patch?

Undef-a replied:
You are correct. This was a mobian specific change. I'll remove it.

ollieparanoid commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -83,8 +85,22 @@ PartitionJob::exec()
+    if ( m_fstype == QString("f2fs") )
+    {
+        m_cmdMkfsRoot.replace(QString("-L"), QString("-l"));
+    }
```
Instead of assuming that cmdMkfsRoot has `@@FSTYPE@@` and replacing that, and replacing `-L` with `-l` in the f2fs case, let's add separate config options for each filesystem type.

Please create a separate commit, which renames `cmdMkfsRoot` to `cmdMkfsRootExt4`. Then add `cmdMkfsBtrfs` and `cmdMkfsF2fs` in your patch and use it here.

(Not many distros have adopted the mobile module yet, so in my opinion we can just rename `cmdMkfsRoot` to `cmdMkfsRootExt4`, without providing backwards compatibility.)

ollieparanoid commented on modules/mobile/mobile.qml:
```diff
@@ -36,6 +36,8 @@ Page
         {"name": "sshd",
          "screens": ["ssh_confirm", "ssh_credentials"]},
+		{"name": "fstype",
+		 "screens": ["fs_selection"]},
```
tabs used instead of spaces

ollieparanoid commented on modules/mobile/mobile.qrc:
```diff
@@ -10,6 +10,7 @@
   <file>default_pin.qml</file> <!-- default user: pin -->
   <file>ssh_confirm.qml</file> <!-- sshd: enable or not? -->
   <file>ssh_credentials.qml</file> <!-- sshd user: username, password -->
+  <file>fs_selection.qml</file> <!-- filesystem selection -->
```
`fs_selection.qml` is missing from the PR?

Undef-a replied:
Yes. I must have missed it trying to remove Mobian specific code. Added now.

ollieparanoid commented on modules/mobile/mobile.qml:
```diff
@@ -36,6 +36,8 @@ Page
          "screens": ["default_pin"]},
         {"name": "sshd",
          "screens": ["ssh_confirm", "ssh_credentials"]},
+		{"name": "fstype",
```
Please add a `featureFstype` config option, so distributions can choose to disable this feature (again, see 7f750cb1450f133c6bc505e60192804b46ff594d).

I would set it to `false` by default, so there are less paths to test when using the `mobile` module with a new Linux distribution.

ollieparanoid commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -83,8 +85,22 @@ PartitionJob::exec()
 
+	// BTRFS requires some extra manipulation for submodules/snapshots
+	if (m_fstype == QString("btrfs")) {
+		m_cmdMkfsRoot.replace(QString("@@FSTYPE@@"), QString("btrfs-installer"));
```
From the commit message:
> OS Maintainers are required to provide their own BTRFS install script
> which configures default submodules as required.

As written in the other comment, I'd put the btrfs command into `cmdMkfsRootBtrfs`. I'd put `mkfs.btrfs` there by default in the calamares-extensions.git repository.

The installer OS could configure the submodules then either by changing this to a `btrfs-installer` script, or by configuring the submodules in a `shellprocess` install step (call a script that checks if the target partition is btrfs, and if so, configure the modules). You can mention something along these lines in the description of the `cmdMkfsRootBtrfs` option in `mobile.conf`.

ollieparanoid reviewed (changes requested):
Thanks for making the PR, @Undef-a. This is a pretty useful feature!

Here goes my review.

Undef-a commented:
Thanks for the review @ollieparanoid. Plenty of good points there.

ollieparanoid commented on modules/mobile/Config.cpp:
```diff
@@ -27,10 +27,13 @@ Config::setConfigurationMap( const QVariantMap& cfgMap )
-    m_cmdMkfsRoot = getString( cfgMap, "cmdMkfsRoot", "mkfs.ext4 -L 'unknownOS_root'" );
+    m_cmdMkfsRootExt4 = getString( cfgMap, "cmdMkfsRootExt4", "mkfs.ext4 -L 'unknownOS_root'" );
+    m_cmdMkfsRootF2fs = getString( cfgMap, "cmdMkfsRootF2fs", "mkfs.f2fs -l 'unknownOS_root'" );
+    m_cmdMkfsRootBtrfs = getString( cfgMap, "cmdMkfsRootBtrfs", "mkfs.btrfs -L 'unknownOS_root'" );
```
Please order them alphabetically (btrfs, ext4, f2fs). Here and the several times they were added below.

ollieparanoid commented on modules/mobile/Config.h:
```diff
@@ -31,12 +31,18 @@ class Config : public QObject
 
+    /* filesystem selection */
+    Q_PROPERTY( QString fstype READ fstype WRITE setFsType NOTIFY fsTypeChanged )
+	Q_PROPERTY( bool featureFstype READ featureFstype CONSTANT FINAL )
```
Indentation is wrong here, and also several times throughout the file.

Run this on each commit:
```sh
clang-format -i $(find . -regex '.*\.\(c\|cpp\|h\)')
```

(Existing commits need to be squashed into one though, from a quick glance they are all just partially implementing the feature and wouldn't work on their own.)

ollieparanoid commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -83,8 +89,17 @@ PartitionJob::exec()
+    QString m_cmdMkfsRoot = m_cmdMkfsRootExt4;
     QList< QPair< QStringList, QString > > commands = {};
 
+    if (m_fstype == QString("btrfs")) {
```
switch ... case seems more suitable here.

ollieparanoid commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -83,8 +89,17 @@ PartitionJob::exec()
+        m_cmdMkfsRoot = m_cmdMkfsRootF2fs;
+    }
+
+
```
(unrelated new empty lines)

ollieparanoid commented on modules/mobile/fs_selection.qml:
```diff
@@ -0,0 +1,56 @@
+    height: parent.height
+
+    Text {
+        id: welcomeText
```
change to `mainText` (see https://github.com/calamares/calamares-extensions/commit/a6a9dc68902327c51aca47bb3eaa42758d6b2bd0)

ollieparanoid commented on modules/mobile/fs_selection.qml:
```diff
@@ -0,0 +1,56 @@
+	ComboBox {
+		id: fstype
+		anchors.horizontalCenter: parent.horizontalCenter
+		anchors.top: welcomeText.bottom
```
`mainText.bottom`

ollieparanoid commented on modules/mobile/Config.h:
```diff
@@ -79,13 +85,20 @@ class Config : public QObject
     QString cmdLuksOpen() const { return m_cmdLuksOpen; }
-    QString cmdMkfsRoot() const { return m_cmdMkfsRoot; }
+    QString cmdMkfsRootExt4() const { return m_cmdMkfsRootExt4; }
+    QString cmdMkfsRootF2fs() const { return m_cmdMkfsRootf2fs; }
```
```suggestion
    QString cmdMkfsRootF2fs() const { return m_cmdMkfsRootF2fs; }
```

ollieparanoid commented on modules/mobile/mobile.conf:
```diff
@@ -69,7 +73,12 @@
+# cmdMkfsRootF2fs: "mkfs.f2fs -l 'unknownOS_root'"
+#
+## To allow snapshots to work on the root subvolume, it is recommended that this
+## command be a script which will create a subvolume and make it default
```
Can you add a link here to the script you are using in mobian, for reference?

Undef-a replied:
I can link to the PR, but not the final script which will be merged in after this. It should be changed when the Mobian release is made.

ollieparanoid replied:
Thanks, this is very useful!

ollieparanoid commented on modules/mobile/fs_selection.qml:
```diff
@@ -0,0 +1,56 @@
+        anchors.topMargin: 40
+        width: 500
+
+        text: qsTr("OK")
```
Other screens that don't have a selection, have "Continue" as button text.

ollieparanoid commented on modules/mobile/mobile.qml:
```diff
@@ -36,6 +36,8 @@ Page
         {"name": "sshd",
          "screens": ["ssh_confirm", "ssh_credentials"]},
+        {"name": "fstype",
+         "screens": ["fs_selection"]},
```
the title for `fs_selection` needs to be set in `titles` above, right now it says "undefined"

ollieparanoid commented on modules/mobile/fs_selection.qml:
```diff
@@ -0,0 +1,56 @@
+		height: 60
+		editable: false
+		model: ["ext4", "f2fs", "btrfs"]
+	}
```
`config.fstype` needs to be selected by default, so:
* the default fstype from the `mobile.conf` is selected first
* when going back to this screen, the previously selected value is still selected. Right now it forgets what was selected if you click OK and then go back. (all other screens show previous input when going back.)


Apparently we need to set `currentIndex` for that, here is an example: https://doc.qt.io/qt-5/qml-qtquick-controls2-combobox.html#combobox-model-roles (note that currentIndex takes an integer).

ollieparanoid commented on modules/mobile/fs_selection.qml:
```diff
@@ -0,0 +1,56 @@
+        anchors.topMargin: 30
+        wrapMode: Text.WordWrap
+
+        text: "Select filesystem for root partition."
```
I've tried to word the other screens with casual users in mind, so:

```suggestion
        text: "Select the filesystem for root partition. If unsure, leave the default."
```

(I think the installer would benefit from a screen at the beginning, that asks if the user wants an advanced mode or a simple mode, the simple mode would then hide the ssh question and the filesystem question and whatever power user stuff we add in the future; but that's out of scope for this PR.)

ollieparanoid commented on modules/mobile/mobile.conf:
```diff
@@ -53,6 +53,10 @@
 
+## Ask user filesystem to install from. If enabled, the user will be able to
+# select between ext4 (default), f2fs and btrfs.
+# featureFstype: false
```
as written in the other comment, the default fs type needs to be configurable here.

Undef-a commented:
Currently this has a bug which causes the external to internal installer to be invoked when not configured.

ollieparanoid commented on modules/mobile/Config.cpp:
```diff
@@ -27,14 +28,24 @@ Config::setConfigurationMap( const QVariantMap& cfgMap )
+    m_cmdMkfsRootBtrfs = getString( cfgMap, "cmdMkfsRootBtrfs", "mkfs.btrfs -L 'unknownOS_root'" );
+    m_cmdMkfsRootExt4 = getString( cfgMap, "cmdMkfsRootExt4", "mkfs.ext4 -L 'unknownOS_root'" );
+    m_cmdMkfsRootF2fs = getString( cfgMap, "cmdMkfsRootF2fs", "mkfs.f2fs -l 'unknownOS_root'" );
+    m_fsmodel = getString( cfgMap, "fsmodel", "ext4,f2fs,btrfs" );
```
For consistency: `fsModel` instead of `fsmodel` (throughout whole patch)

ollieparanoid commented on modules/mobile/Config.cpp:
```diff
@@ -27,14 +28,24 @@ Config::setConfigurationMap( const QVariantMap& cfgMap )
+    m_fsmodel = getString( cfgMap, "fsmodel", "ext4,f2fs,btrfs" );
+    m_fsLabelModel = new Locale::LabelModel( m_fsmodel.split( ',' ) );
+    m_defaultFs = getString( cfgMap, "defaultFs", "ext4" );
+    /* Locale::Label l_defaultLocale = new Locale::Label( m_defaultFs, Locale::Label::LabelFormat::IfNeededWithCountry, m_fsLabelModel ); */
```
This line of commented out code looks like a left-over. Remove it?

(Are you working on a patch for locale support too? Or was this used as template from some calamares code?)

Undef-a replied:
That was just a left over from trying to work out the Locale bit of the LabelModel. It'll be gone shortly.

ollieparanoid commented on modules/mobile/Config.cpp:
```diff
@@ -143,3 +157,29 @@ Config::setInstallFromExternalToInternal( const bool val )
+    emit fsTypeChanged( m_fsType );
+}
+void
+Config::setFsType( const QString& fstype )
```
`fstype` -> `fsType` (throughout the whole patch)

ollieparanoid commented on modules/mobile/Config.h:
```diff
@@ -79,13 +89,25 @@ class Config : public QObject
     void setIsFdeEnabled( bool isFdeEnabled );
 
+    /* filesystem selection */
+    bool featureFstype() { return m_featureFstype; };
```
`featureFsType`

ollieparanoid commented on modules/mobile/Config.h:
```diff
@@ -120,11 +142,21 @@ class Config : public QObject
+    QString m_fsType = m_defaultFs;
+    CalamaresUtils::Locale::LabelModel* m_fsLabelModel;
+    int m_fsIndex;
+    QString m_fsmodel = "ext4,f2fs,btrfs";
```
I suggest we put the default values in `setConfigurationMap()` instead of duplicating them here, because:

* One less place to adjust if we want to change the default.
* Consistency with the other config options above.

Therefore I'd remove the default string from `m_fsmodel` and `m_defaultFs`.

ollieparanoid commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -83,8 +89,22 @@ PartitionJob::exec()
+    {
+        m_cmdMkfsRoot = m_cmdMkfsRootF2fs;
+    }
+    else
```
To prevent unintended behaviour, let's explicitly check for ext4 here, and throw an error if the fstype is anything else (see ` return JobResult::error` below).

ollieparanoid commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -83,8 +89,22 @@ PartitionJob::exec()
     QString cryptDev = "/dev/mapper/" + cryptName;
     QString passwordStdin = m_password + "\n";
     QString dev = m_targetDeviceRoot;
+    QString m_cmdMkfsRoot = m_cmdMkfsRootExt4;
```
Whatever we set here is overwritten below, so let's change this to `QString m_cmdMkfsRoot;`.

ollieparanoid commented on modules/mobile/mobile.conf:
```diff
@@ -53,6 +53,12 @@
+# select between ext4 (default), f2fs and btrfs.
+# featureFstype: false
+# fsmodel: ext4,f2fs,btrfs
+# defaultFs: ext4
```
```suggestion
## Ask the user, which filesystem to use.
# featureFsType: false
## Filesystems that the user can choose from.
# fsModel: ext4,f2fs,btrfs
## Default filesystem to display in the dialog. If featureFsType is disabled,
## this gets used without asking the user.
# defaultFs: ext4
```

changes:
* explained each option.
* explicitly wrote that defaultFs gets used even if featureFsType is false, as this might not be obvious.
* changed featureFstype -> featureFsType and fsmodel -> fsModel (see other comments).

ollieparanoid commented on modules/mobile/mobile.conf:
```diff
@@ -69,7 +75,14 @@
+## An example can be found at:
+## https://gitlab.com/mobian1/calamares-settings-mobian/-/merge_requests/2/diffs#diff-content-dde34f5f1c89e3dea63608c553bbc452dedf428f
+# cmdMkfsRootBtrfs: "mkfs.btrfs -L 'unknownOS_root'"
+#
```
I suggest to remove this line, and start the btrfs related comment above with `Btrfs: to allow snapshots...`.

ollieparanoid commented on modules/mobile/mobile.conf:
```diff
@@ -69,7 +75,14 @@
 
 ## Format the rootfs with a file system
 ## Arguments: <device>
-# cmdMkfsRoot: "mkfs.ext4 -L 'unknownOS_root'"
```
Can you mention in https://github.com/calamares/calamares-extensions/blob/calamares/CHANGES, that `cmdMkfsRoot` was renamed to `cmdMkfsRootExt4`?

ollieparanoid commented on modules/mobile/fs_selection.qml:
```diff
@@ -0,0 +1,62 @@
+        textRole: "label"
+        /* Save the current state on selection so it is there when the back button is pressed */
+        onActivated: config.fsType = fstypeCB.currentText.split(" ")[1];
+        /* Label we are searching for is of the format "* ext4 (C)" */
```
`* ext4 (C)` looks pretty weird in the UI.

@adriaandegroot: can you give a hint how we could set a label text?

(If this is much effort, I suggest to do that in a future PR and rather get this one merged soon.)

adriaandegroot replied:
Use a model for string lists, rather than one for selecting locales; even better is probably writing a model (it would go in `libcalamares/partition/FileSystem.h` most likely) that shows a handful of filesystem names / types.

ollieparanoid reviewed (changes requested):
Great work, @Undef-a! You even figured out how to get the combobox entries dynamically generated and how to make it properly display the previous selection when using the back button. :tada: 

I've tried your patch and it works as expected.

I think this is almost ready to merge now, review is mostly consistency nitpicks.

---

EDIT: 

> Currently this has a bug which causes the external to internal installer to be invoked when not configured.

I'll look into that.

EDIT2: cannot reproduce. If `targetDeviceRootInternal` is not in `mobile.conf`, the feature will be skipped (see [`skipFeatureInstallTarget()`](https://github.com/calamares/calamares-extensions/blob/a53f9359f42b6c301b19060f90a5d4e8bc771977/modules/mobile/mobile.qml#L150-L152)). I just tried it out and it got skipped as expected:
```
07:46:37 [8]: Navigating to screen: welcome
07:46:39 [8]: Skipping feature (skip function): installTarget
07:46:39 [8]: Navigating to feature: userPin
07:46:39 [8]: Navigating to screen: default_pin
```
If you still have the problem, please open an issue here in calamares-extensions and mention me.

Undef-a commented:
> cannot reproduce. If `targetDeviceRootInternal` is not in `mobile.conf`, the feature will be skipped (see [`skipFeatureInstallTarget()`](https://github.com/calamares/calamares-extensions/blob/a53f9359f42b6c301b19060f90a5d4e8bc771977/modules/mobile/mobile.qml#L150-L152)). I just tried it out and it got skipped as expected:

That comment was made as a not for my on testing. It seemed to happen a few times when I was recently trying to flash my phone. If it's not reproducible, its probably related to the semi-finished Mobian packaging I was using.

ollieparanoid commented:
Looks good to me! :+1: Thanks again for all the work you've put into this, @Undef-a!

@adriaandegroot: can you please take a look at this?

If you think it's ready to merge, it would be great to merge #10 as well and tag a new release.

adriaandegroot commented on modules/mobile/Config.h:
```diff
@@ -79,13 +89,25 @@ class Config : public QObject
+    QString fsType() const { return m_fsType; };
+    void setFsType( const int idx );
+    void setFsType( const QString& fsType );
+    CalamaresUtils::Locale::LabelModel* fsModel() const;
```
This is a really peculiar choice. `Locale::LabelModel` is a model containing names of locales: it offers current-language and native-language names for locales (e.g. "Dutch" and "Nederlands"). If a locale is unknown to Qt (e.g. Esperanto, before Qt 5.12) then an additional "(C)" is added, indicating that Calamares is going to treat this as C locale (e.g. no translations, no locale support).

I suppose you **can** stick a handful of strings in it, but it does not seem like a good fit. For what it's worth, KPMCore offers a list of (translated) filesystem names.

adriaandegroot commented on modules/mobile/Config.h:
```diff
@@ -79,13 +89,25 @@ class Config : public QObject
+    /* filesystem selection */
+    bool featureFsType() { return m_featureFsType; };
+    QString fsType() const { return m_fsType; };
+    void setFsType( const int idx );
```
`int` is a value type, there's not much point in making it `const` (but that's C++-stylistic nitpicking)

adriaandegroot commented on modules/mobile/Config.h:
```diff
@@ -31,12 +32,21 @@ class Config : public QObject
-    Q_PROPERTY( QString cmdMkfsRoot READ cmdMkfsRoot CONSTANT FINAL )
+    Q_PROPERTY( QString cmdMkfsRootBtrfs READ cmdMkfsRootBtrfs CONSTANT FINAL )
+    Q_PROPERTY( QString cmdMkfsRootExt4 READ cmdMkfsRootExt4 CONSTANT FINAL )
+    Q_PROPERTY( QString cmdMkfsRootF2fs READ cmdMkfsRootF2fs CONSTANT FINAL )
```
Do these really need to be exposed as properties? Seems like an internal thing to me (so was the previous `cmdMkfsRoot`). Unless you're expecting these to be displayed or edited from QML, it's better to hide them.

adriaandegroot commented on modules/mobile/Config.h:
```diff
@@ -120,11 +142,21 @@ class Config : public QObject
+    bool m_featureFsType;
+    QString m_defaultFs;
+    QString m_fsType;
+    CalamaresUtils::Locale::LabelModel* m_fsLabelModel;
```
See above -- LabelModel is intended to store names of locales (and their translations), not arbitrary string lists. It's a lousy name, I'll give you that (and there's no documentation for that class).

adriaandegroot commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -83,8 +89,26 @@ PartitionJob::exec()
     QString cryptDev = "/dev/mapper/" + cryptName;
     QString passwordStdin = m_password + "\n";
     QString dev = m_targetDeviceRoot;
+    QString m_cmdMkfsRoot;
```
Generally in Calamares code, we use `m_` as a prefix for member variables; this is just a local variable, so it's weird to see it named like that.

adriaandegroot commented on modules/mobile/PartitionJob.cpp:
```diff
@@ -15,24 +15,30 @@
     , m_installFromExternalToInternal( installFromExternalToInternal )
     , m_isFdeEnabled( isFdeEnabled )
     , m_password( password )
+    , m_fsType( fsType )
```
Can `m_fsType` change after the Job is created? If not, why not pass the appropriate command in the constructor (along with the type, for whatever reasons) and leave the logic of picking the right command to the caller / creator of the job?

adriaandegroot commented on modules/mobile/mobile.conf:
```diff
@@ -53,6 +53,14 @@
+## Ask the user, which filesystem to use.
+# featureFsType: false
+## Filesystems that the user can choose from.
+# fsModel: ext4,f2fs,btrfs
```
If you really, specifically, explicitly want to use comma-separated identifiers here, you should document that. Otherwise it's probably nicer (although it doesn't save any code elsewhere) to specify this as a YAML list.

adriaandegroot commented:
Lots of comments, but the decision is down to you: if it works for your use-case and gives you the UI you want, then it's fine.

ollieparanoid commented:
Thanks a lot for the thorough review! I think these review points should be addressed before we merge it.

> Use a model for string lists, rather than one for selecting locales; even better is probably writing a model (it would go in libcalamares/partition/FileSystem.h most likely) that shows a handful of filesystem names / types.

Getting the file system selection right seems like the most complex part... @adriaandegroot, can you write a draft of how the addition to FileSystem.h should look like?

Undef-a commented:
I think all of the above are now resolved, thanks for the through review. I need to build/flash a test image of this on an actual phone, but pending that I believe this is ready for another review.

adriaandegroot commented on modules/mobile/Config.cpp:
```diff
@@ -27,14 +27,22 @@ Config::setConfigurationMap( const QVariantMap& cfgMap )
     m_targetDeviceRoot = getString( cfgMap, "targetDeviceRoot", "/dev/unknown" );
     m_targetDeviceRootInternal = getString( cfgMap, "targetDeviceRootInternal", "" );
 
+    m_cmdMkfsRootBtrfs = getString( cfgMap, "cmdMkfsRootBtrfs", "mkfs.btrfs -L 'unknownOS_root'" );
```
Are distro's expected to set the configuration values (in `mobile.conf`) in order to change the FS label? Perhaps it would be useful to define some substitutions and then use KMacroExpander .. but that's for another iteration of this module, let's get this in first.

Suggested substitutions:
- *fslabel*, defined in the configuration file for this module
- *shortname*, which is the branding-defined short name of the distro
- *root*, which expands to .. "root" I guess

Then the make-fs commands could be defined as `mkfs.ext4 -L '${fslabel}'`, for instance, and if a distro wants to change that, only changing *fslabel* in the config file (once!) is needed, rather than updating three commands.

But: future work

ollieparanoid replied:
That could work, but yes, out of scope for this PR.

To explain a bit why the full FS command can be set: distros may have specific options for the FS commands (for example in postmarketOS we have two ext4 options to ensure that it works with older downstream kernels too), or use different commands altogether (e.g. for btrfs, which needs to be wrapped in a script that also creates subvolumes and sets default ones).

adriaandegroot commented on modules/mobile/Config.cpp:
```diff
@@ -66,6 +74,24 @@ Config::createJobs()
+    else if ( m_fsType == QStringLiteral( "ext4" ) )
+    {
+        cmdMkfsRoot = m_cmdMkfsRootExt4;
+    }
```
I'm not *really* thrilled with string comparisons like this, since it gives opportunities for typo's and other things that are caught at runtime, rather than compile-time. I don't immediately see how to improve this, though, since introducing an enum becomes complicated with the QML parts that use a string list for display (or you'd have to move to a more expressive model).

adriaandegroot commented on modules/mobile/Config.cpp:
```diff
@@ -66,6 +74,24 @@ Config::createJobs()
+    }
+    else
+    {
+        v->onInstallationFailed( "Unknown filesystem: '" + m_fsType + "'", "");
```
Clever! This is needed, of course, since you're running jobs here which kind of goes against the grain of how Calamares is designed, but it **does** let you report failure effectively.

Note that you're missing a `tr()` call here, so it is not going to be translated (but, come to that, translations of calamares-extensions are not available at all).

adriaandegroot commented on modules/mobile/Config.cpp:
```diff
@@ -78,16 +104,15 @@ Config::runPartitionJobThenLeave( bool b )
                                           m_installFromExternalToInternal,
                                           m_isFdeEnabled,
-                                          m_fdePassword );
+                                          m_fdePassword);
```
Calamares coding style says "put that space there", but that's an easy fix with the styling-tool.

adriaandegroot commented on modules/mobile/Config.h:
```diff
@@ -79,13 +85,25 @@ class Config : public QObject
+    QString fsType() const { return m_fsType; };
+    void setFsType( int idx );
+    void setFsType( const QString& fsType );
+    QStringList fsList() const;
```
Why is this getter not inline? It doesn't matter much, but it would be stylistically consistent to also `{return m_fsList; }` here, and not have the out-of-line definition in `Config.cpp`.

adriaandegroot commented on modules/mobile/Config.cpp:
```diff
@@ -143,3 +168,29 @@ Config::setInstallFromExternalToInternal( const bool val )
+void
+Config::setFsType( int idx )
+{
+    m_fsType = m_fsList[ idx ];
```
This body could be squashed down to `setFsType( m_fsList[ idx ] );` .. doesn't save much, but gives you **one** place to check for validity and emit related signals.

adriaandegroot replied:
Do you think bounds-checking `idx` is needed? It seems slightly hazardous so accept any int here; if the QML code is changed (I dunno, let's set the type to -1 while we don't know what to use) then things will go very wrong here.

adriaandegroot commented on modules/mobile/Config.cpp:
```diff
@@ -143,3 +168,29 @@ Config::setInstallFromExternalToInternal( const bool val )
+void
+Config::setFsType( const QString& fsType )
+{
+    m_fsType = fsType;
```
Often setters will check if the change is actually a change, to avoid sending useless signals (e.g. add a `if (fsType != m_fsType)` block)

adriaandegroot commented on modules/mobile/Config.h:
```diff
@@ -120,11 +138,20 @@ class Config : public QObject
+    bool m_featureFsType;
+    QString m_defaultFs;
+    QString m_fsType;
+    int m_fsIndex;
```
What is the index for? It looks like "just a number" and I don't see it being used. Maybe some documentation in the `Config.h` header -- say on the getters for these properties -- would help.

adriaandegroot commented:
So, I'm going to try to set up CI for calamares-extensions first, then merge this, then apply coding-style and minor tweaks to the code. The comments I have just made in the review are "this would be a neat extra feature" kind of things, and trivia like coding-style. Overall it looks pretty solid and warrants a new release of calamares-extensions.

adriaandegroot commented:
@Undef-a CI is now set up, just give me (or @ollieparanoid , who can then ping me on IRC) a sign if you think it can be merged; I don't want to do that if you're still working on this branch because it would mess up your workflow.

Undef-a commented:
After that squash and last config change I'm pretty happy with it, and I can work on the polish you brought up in another branch. Just writing an image to test it on a real device now. 

I'll yell out in #calamares when that is done.

Undef-a commented:
My testing is triggering the internal from external installer again. This is using the mobian packaging, which currently does not have any logic for this function. 

It also happens on the stock mobile.conf from this PR. 

Until I can work out why this is happening, this probably should not be merged.

ollieparanoid commented:
@Undef-a: weird, I'll ping you in IRC so we can work this out.

Undef-a commented:
Successfully tested on device. Lucky I did as there were two blocking bugs:

- I changed the order of arguments in PartitionJob. 
- External -> internal install was triggered. 

@ollieparanoid Thanks for the help with these. Could you please review the fix for the external -> internal issue. I believe that when the option is not triggered, this was defaulting to "true" on Mobian, resulting in the error. I have initialized this variable to fix it, which solved the issue, but I can't confirm that it won't cause you issues.

ollieparanoid commented on modules/mobile/Config.h:
```diff
@@ -120,15 +138,25 @@ class Config : public QObject
     QString m_targetDeviceRoot;
     QString m_targetDeviceRootInternal;
-    bool m_installFromExternalToInternal;
+    bool m_installFromExternalToInternal = false;
```
I just read the code again where the variable gets used, and indeed, we must set this to false!

Otherwise it is uninitialized if the install target screen is skipped. Looks like it was only "false" by chance in my testing, and this went unnoticed in the proper postmarketOS installer images because we don't skip the dialog there (unless flashing to eMMC, which most people don't do).

Can you put this bugfix into a separate commit inside this PR?

ollieparanoid reviewed (approved):
> Could you please review the fix for the external -> internal issue. I believe that when the option is not triggered, this was defaulting to "true" on Mobian, resulting in the error. I have initialized this variable to fix it, which solved the issue, but I can't confirm that it won't cause you issues.
> 
> EDIT: Specifically referring to L159 of Config.h

Tested with your branch with both installing to internal and external storage with QEMU, and it still works as expected. Moreover you found and fixed a real bug there, the variable was uninitialized and I didn't notice.

This PR is ready to merge in my opinion, only the commits need to be squashed properly, and the bugfix should go into an own commit.

Let's ship this! :ship:  Thank you so much for your hard work on this, @Undef-a and for all the reviewing, @adriaandegroot!

PS: regarding translations: I think we should add tr() around all the strings eventually, and use template strings in javascript to fill in the variables (at least the QT version used in pmOS supports this, we should also check if it works with debian's QT). But of course, that's for another PR.

Commit:
[mobile] bugfix: initialise externalToInternal

This variable was uninitialised, leading to an undefined action occurring
when the install target screen is skipped.
